groups/bsl/bslstl/bslstl_ratio.t.cpp

```diff
@@ -0,0 +1,349 @@
+
+        if (verbose) printf("\nTesting Usage Example"
+                            "\n=====================\n");
+
+      } break;
+      case 1: {
+        // ----------------------------------------------------------------------
+        // TESTING the ratio operations
+        //
+        // Concerns:
+        //: 1 That the operations are successfully imported into the 'bsl'
+        //:   namespace from the underlying STL implementation.
+        //: 
+        //: 2 That the short versions of the comparison operators, such
+        //:   as 'ratio_less_v' exist and work even if the underlying STL
+        //:   does not provide them.
+        //
+        // Plan:
+        //: 1 Verify that the SI typedefs exist and have the correct 
+        //:   numerator and denominator.
+        //:
+        //: 2 Verify that the comparison operations exist and give 
+        //:   expected results. Check again for the '_v' versions.
+        //:
+        //: 3 Verify that the arithmetic operations exist and give
+        //:   expected results.
+        //
+        // ----------------------------------------------------------------------
+
+        if (verbose) printf("\nTesting ratio operations"
+                            "\n======================================\n");
+
+        if (verbose) printf("\tTest SI unit prefixes.\n");
+        {
+            ASSERT(1 == bsl::deci::num);
+            ASSERT(1 == bsl::centi::num);
+            ASSERT(1 == bsl::milli::num);
+            ASSERT(1 == bsl::micro::num);
+            ASSERT(1 == bsl::nano::num);
+            ASSERT(1 == bsl::pico::num);
+            ASSERT(1 == bsl::femto::num);
+            ASSERT(1 == bsl::atto::num);
+
+            ASSERT(10ULL == bsl::deci::den);
+            ASSERT(100ULL == bsl::centi::den);
+            ASSERT(1000ULL == bsl::milli::den);
+            ASSERT(1000000ULL == bsl::micro::den);
+            ASSERT(1000000000ULL == bsl::nano::den);
+            ASSERT(1000000000000ULL == bsl::pico::den);
+            ASSERT(1000000000000000ULL == bsl::femto::den);
+            ASSERT(1000000000000000000ULL == bsl::atto::den);
+
+            ASSERT(1 == bsl::deca::den);
+            ASSERT(1 == bsl::hecto::den);
+            ASSERT(1 == bsl::kilo::den);
+            ASSERT(1 == bsl::mega::den);
+            ASSERT(1 == bsl::giga::den);
+            ASSERT(1 == bsl::tera::den);
+            ASSERT(1 == bsl::peta::den);
+            ASSERT(1 == bsl::exa::den);
+
+            ASSERT(10ULL == bsl::deca::num);
+            ASSERT(100ULL == bsl::hecto::num);
+            ASSERT(1000ULL == bsl::kilo::num);
+            ASSERT(1000000ULL == bsl::mega::num);
+            ASSERT(1000000000ULL == bsl::giga::num);
+            ASSERT(1000000000000ULL == bsl::tera::num);
+            ASSERT(1000000000000000ULL == bsl::peta::num);
+            ASSERT(1000000000000000000ULL == bsl::exa::num);
+
+        #ifdef BSL_RATIO_SUPPORTS_EXTENDED_SI_TYPEDEFS
```

Comment:
This macro needs to be defined in the header file (email to Alisdair).  I think it would be nice to resolve this issue in this PR also (depending on what Alisdair says).

Reply:
I've done some investigation about this in the last couple days. No standard library that I know of implements these; because you need `sizeof(intmax_t) > 8` - and that leaves out all the platforms that I'm aware of.